model_analyzer/triton/client/client.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import time
+
+from tritonclient.utils import InferenceServerException
+from model_analyzer.model_analyzer_exceptions import TritonModelAnalyzerException
+
+WAIT_FOR_READY_NUM_RETRIES = 100
+
+
+class TritonClient:
+    """
+    Defines the interface for the objects created by 
+    TritonClientFactory
+    """
+
+    def wait_for_server_ready(self, num_retries=WAIT_FOR_READY_NUM_RETRIES):
+        """
+        Parameters
+        ----------
+        num_retries : int
+            number of times to send a ready status
+            request to the server before raising
+            an exception
+        
+        Raises
+        ------
+        TritonModelAnalyzerException
+            If server readiness could not be
+            determined in given num_retries
+        """
+        while num_retries > 0:
+            try:
+                if self._client.is_server_ready():
+                    return
+                else:
+                    time.sleep(0.1)
+                    num_retries -= 1
+            except Exception as e:
+                time.sleep(0.1)
+                num_retries -= 1
+                pass
+        raise TritonModelAnalyzerException("Could not determine server readiness. "
+                                           "Number of retries exceeded.")
+    
+    def load_model(self, model_name):
```

Comment:
Ashwin, I was looking at the PR again. I was wondering what do you think about adding a `TritonModel` or `Model` class that encapsulates all the information related to a Triton Model and using that instead of only relying on the `model_name` string.

Reply:
I think you're right. For future phases, I'm sure we'll have to create either a `TritonModel` or a `TritonModelConfig` class that encapsulates the information. Do you think the load model, and unload model interface should belong to `TritonModel`?

Reply:
Sorry for the delayed response. I think it would be great to have `load` and `unload` as a part of the model class. But I guess the problem will be we have to pass a client to every TritonModel. I guess we can probably create a TritonClient in every model or make the TritonClient singleton (if it doesn't need to be instantiated) so that in the model constructor we used the same instance all the time.

Reply:
Hmm I suppose what we could have is a singleton class for each concrete implementation of the TritonClient? So there is one GRPC client and one HTTP client for each? Now I'm wondering whether we even need a standalone TritonClient interface, or whether we'll only ever need to interact via a `TritonModel` interface.

Reply:
In the C# version of the Model Analyzer, they only used TritonClient to `load` models into the Triton. Maybe in the future (esp. with the introduction of rate limiter) we may need TritonClient for setting up the general way that Triton works independently from the models. So I think it would be better to keep it as a separate entity. The other thing is that we will have a single type of TritonClient (either gRPC or HTTP) throughout the application execution. So if it is instantiated once (using gRPC or HTTP), the same instance can be used throughout the application lifetime.

Reply:
You don't want a singleton TritonClient. That would imply that both HTTPClient and GRPClient exist at the same time. A singleton object is the same a global object. Why would everything need to know about both HTTPClient and GRPClient? 

In this case, you want a top level Model Interface that defines the load/unload and other common methods. Then a derived class for HTTP Client and GRPC Client. You can then use a Constructor Pattern to create them. However, unless there is a use case, both don't need to exist at the same time. You will use one or the other.

Reply:
FYI, We will definitely need both a Model and ModelConfig object. I don't think you need to specify Triton exactly unless we are analyzing models for a non Triton platform. The Model object should hold all the meta data about the model, while the ModelConfig holds the configuration options for a Model under test. You may have an array of ModelConfigs, one for each test. Then when we create the actual Triton ModelConfig, the writer would take both as input and write out the correct config file.

Reply:
But also don't plan for what ifs right now. Instead, create the minimal amount of code needed to pass the tests and use cases for this phase. Keep the code agile and clean. Then we can refactor as needed when more use cases are added.